app/services/fact_importer.rb

```diff
@@ -39,31 +39,86 @@ def initialize(host, facts = {})
 
   # expect a facts hash
   def import!
+    ensure_fact_names
+
     ActiveRecord::Base.transaction do
       delete_removed_facts
       update_facts
       add_new_facts
     end
 
-    if @error
-      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
-      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s"), @host.name)
-    end
+    report_error(@error) if @error
     logger.info("Import facts for '#{host}' completed. Added: #{counters[:added]}, Updated: #{counters[:updated]}, Deleted #{counters[:deleted]} facts")
   end
 
+  def report_error(error)
+    Foreman::Logging.exception("Error during fact import for #{@host.name}", error)
+    raise ::Foreman::WrappedException.new(error, N_("Import of facts failed for host %s"), @host.name)
+  end
+
   # to be defined in children
   def fact_name_class
     raise NotImplementedError
   end
 
   private
 
-  attr_reader :host, :facts
+  attr_reader :host, :facts, :fact_names, :fact_names_by_id
+
+  def ensure_fact_names
+    initialize_fact_names
+
+    missing_keys = facts.keys - fact_names.keys
+    add_missing_fact_names(missing_keys)
+  end
+
+  def add_missing_fact_names(missing_keys)
+    missing_keys.sort.each do |fact_name|
+      # create a new record and make sure it could be saved.
+      name_record = fact_name_class.new(fact_name_attributes(fact_name))
+
+      ActiveRecord::Base.transaction(:requires_new => true) do
+        begin
+          unless name_record.save
+            throw ActiveRecord::RecordInvalid.new(name_record) unless duplicate_name_error?(name_record)
+
+            # if the record could not be saved, then it should already be in the database
+            # throw an exception if it isn't
+            name_record = fact_name_class.find_by!(name: fact_name)
+          end
+        rescue ActiveRecord::RecordNotUnique
+          name_record = fact_name_class.find_by!(name: fact_name)
+        end
+      end
+
+      # make the new record available immediately for other fact_name_attributes calls
+      @fact_names[fact_name] = name_record
+    end
+  end
+
+  def duplicate_name_error?(name_record)
+    name_record.errors.size == 1 && name_record.errors.details[:name].first.try(:[], :error) == :taken
+  end
+
+  def initialize_fact_names
+    name_records = fact_name_class.where(:name => facts.keys, :type => fact_name_class.to_s).to_a
+    @fact_names = {}
+    @fact_names_by_id = {}
+    name_records.each do |record|
+      @fact_names[record.name] = record
+      @fact_names_by_id[record.id] = record
+    end
+  end
```

Comment:
All records will still exist in memory in the end of the loop - they are referenced by both hashes.

Reply:
But you can still avoid calling `to_a` it will create a useless array of pointers for no reason.

Anyway, I insist on not creating expensive AR objects here. For `fact_names_by_id` you can only store name, because the only reference you do is `fact_names_by_id[X].name` and there is no reason to store everything.

For `fact_name` you can also avoid storing whole AR and just store ID, but it needs to be proofed by unit tests. You can do this:

```ruby
    fact_name_id = fact_names[name]
    host.fact_values.send(method, :value => facts[name], :fact_name_id => fact_name_id)
```

Another use is here which you can also rewrite to faster:

    host.fact_values.where(:fact_name_id => fact_names.values)

And the last use can be also rewritten to:

    parent_fact_record_id = fact_names[parent_fact_name(name)]
    attributes[:parent_id] = parent_fact_record_id

This will be I believe much faster.